File: src/bin/mt/apksign/data/DataSource.java
package bin.mt.apksign.data;

import bin.io.RandomAccessFile;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.security.MessageDigest;

public interface DataSource {

    long size();

    long pos();

    default long remaining() {
        return size() - pos();
    }

    void reset() throws IOException;

    void copyTo(OutputStream os, long length) throws IOException;

    default void copyTo(MessageDigest digest, long length) throws IOException {
        OutputStream os = new OutputStream() {
            @Override
            public void write(int b) {
                digest.update((byte) b);
            }

            @Override
            public void write(byte[] b, int off, int len) {
                digest.update(b, off, len);
            }
        };
        copyTo(os, length);
    }

    default void copyTo(RandomAccessFile accessFile, long length) throws IOException {
        OutputStream os = new OutputStream() {
            @Override
            public void write(int b) throws IOException {
                accessFile.write(b);
            }

            @Override
            public void write(byte[] b, int off, int len) throws IOException {
                accessFile.write(b, off, len);
            }
        };
        copyTo(os, length);
    }

    default DataSource align(int align) {
        return DataSources.align(this, align);
    }

    default ByteArrayDataSource toMemory() throws IOException {
        long remaining = remaining();
        if (remaining > Integer.MAX_VALUE) {
            throw new IOException("Data too large");
        }
        ByteArrayOutputStream baos = new ByteArrayOutputStream((int) remaining);
        copyTo(baos, remaining);
        return (ByteArrayDataSource) DataSources.fromData(baos.toByteArray());
    }

}

File: Android/app/src/main/java/bin/mt/apksign/data/DataSource.kt
package bin.mt.apksign.data

import bin.io.RandomAccessFile
import bin.mt.apksign.data.DataSources.fromData
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStream
import java.security.MessageDigest

interface DataSource {
    fun size(): Long
    fun pos(): Long
    fun remaining(): Long {
        return size() - pos()
    }

    @Throws(IOException::class)
    fun reset()

    @Throws(IOException::class)
    fun copyTo(os: OutputStream, length: Long)

    @Throws(IOException::class)
    fun copyTo(digest: MessageDigest, length: Long) {
        val os: OutputStream = object : OutputStream() {
            override fun write(b: Int) {
                digest.update(b.toByte())
            }

            override fun write(b: ByteArray, off: Int, len: Int) {
                digest.update(b, off, len)
            }
        }
        copyTo(os, length)
    }

    @Throws(IOException::class)
    fun copyTo(accessFile: RandomAccessFile, length: Long) {
        val os: OutputStream = object : OutputStream() {
            @Throws(IOException::class)
            override fun write(b: Int) {
                accessFile.write(b)
            }

            @Throws(IOException::class)
            override fun write(b: ByteArray, off: Int, len: Int) {
                accessFile.write(b, off, len)
            }
        }
        copyTo(os, length)
    }

    fun align(align: Int): DataSource {
        return DataSources.align(this, align)
    }

    @Throws(IOException::class)
    fun toMemory(): ByteArrayDataSource? {
        val remaining = remaining()
        if (remaining > Int.MAX_VALUE) {
            throw IOException("Data too large")
        }
        val baos = ByteArrayOutputStream(remaining.toInt())
        copyTo(baos, remaining)
        return fromData(baos.toByteArray()) as ByteArrayDataSource?
    }
}